lib/OpenTelemetry/Processor/Batch.pm

```diff
@@ -0,0 +1,195 @@
+        name        => [qw( buffer use )],
+        description => 'Ratio between maximum queue size and the size of the queue at export time',
+    );
+
+    field $batch_size       :param //= config('BSP_MAX_EXPORT_BATCH_SIZE') //    512;
+    field $exporter_timeout :param //= config('BSP_EXPORT_TIMEOUT')        // 30_000;
+    field $max_queue_size   :param //= config('BSP_MAX_QUEUE_SIZE')        //  2_048;
+    field $schedule_delay   :param //= config('BSP_SCHEDULE_DELAY')        //  5_000;
+    field $exporter         :param;
+
+    field $lock = Mutex->new;
+
+    field $done :reader;
+    field $function;
+    field @queue;
+
+    ADJUST {
+        die OpenTelemetry::X->create(
+            Invalid => "Exporter must implement the OpenTelemetry::Exporter interface: " . ( ref $exporter || $exporter )
+        ) unless $exporter && $exporter->DOES('OpenTelemetry::Exporter');
+
+        if ( $batch_size > $max_queue_size ) {
+            $logger->warn(
+                'Max export batch size cannot be greater than maximum queue size when instantiating batch processor',
+                {
+                    batch_size => $batch_size,
+                    queue_size => $max_queue_size,
+                },
+            );
+            $batch_size = $max_queue_size;
+        }
+
+        # This is a non-standard variable, so we make it Perl-specific
+        my $max_workers = $ENV{OTEL_PERL_BSP_MAX_WORKERS};
+
+        $function = IO::Async::Function->new(
+            $max_workers ? ( max_workers => $max_workers ) : (),
+
+            code => sub ( $exporter, $batch, $timeout ) {
+                $exporter->export( $batch, $timeout );
+            },
+        );
+
+        IO::Async::Loop->new->add($function);
+    }
+
+    method process ( @items ) {
+        try {
+            my $batch = $lock->enter(
+                sub {
+                    my $overflow = @queue + @items - $max_queue_size;
+                    if ( $overflow > 0 ) {
+                        # If the buffer is full, we drop old spans first
+                        # The queue is always FIFO, even for dropped spans
+                        # This behaviour is not in the spec, but is
+                        # consistent with the Ruby implementation.
+                        # For context, the Go implementation instead
+                        # blocks until there is room in the buffer.
+                        splice @queue, 0, $overflow;
+                        $self->report_dropped( 'buffer-full', $overflow );
+                    }
+
+                    push @queue, @items;
+
+                    return [] if @queue < $batch_size;
+
+                    $metrics->set_gauge_to(
+                        buffer_use => @queue / $max_queue_size
+                    ) if @queue;
+
+                    [ splice @queue, 0, $batch_size ];
+                }
+            );
+
+            return unless @$batch;
+
+            $function->call(
```

Comment:
This call is not executed in `void` context (I guess it's a side-effect of how `try/catch` is implemented, so it currently creates a future that needs to be waited on.

Reply:
Ooh, that's really very interesting. This should be an easy fix in that case. Thanks!

I'll give it a whirl, and maybe we can unblock this.

Reply:
This worked. Thanks!